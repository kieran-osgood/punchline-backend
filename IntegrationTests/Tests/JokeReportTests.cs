
using System.Text.Json;
using System.Threading.Tasks;
using GraphQL;
using GraphQL.Data;
using GraphQL.Entities.JokeReport;
using HotChocolate;
using HotChocolate.Execution;
using HotChocolate.Types.Relay;
using IntegrationTests.Helpers;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit;

namespace IntegrationTests.Tests
{
    public class JokeReportTests : IClassFixture<CustomWebApplicationFactory<Program>>
    {
        [Fact]
        public async Task Add_Jokes_Report_Commits()
        {
            var factory =
                new CustomWebApplicationFactory<Program>().WithAuthentication(TestClaimsProvider.WithUserClaims());
            var user = Utilities.GetValidAuthUser();
            var joke = Utilities.GetValidJoke();
            const string description = "This joke was despicable!!!!!!!";
            var idSerializer = factory.Services.GetRequiredService<IIdSerializer>();
            var jokeId = idSerializer.Serialize("", nameof(Joke), joke.Id);

            var query = 
$@"mutation RateJoke {{
    addJokeReport(input: {{
        id: {"\""  + jokeId + "\""},
        description: {"\"" + description+ "\""}
    }}) {{
        jokeReport {{
            id
            reportingUser {{
                id
                name
            }}
            reportedJoke {{
                    id
                    title
                }}
        }}
        errors {{
            code
            message
        }}
    }}
}}"; 

            var executor = await factory.Services.GetRequestExecutorAsync();
            var request = QueryRequestBuilder
                .New()
                .AddAuthorizedUser()
                .SetQuery(query)
                .Create();

            var result = await executor.ExecuteAsync(request);
            
            Assert.Null(result.Errors);
            
            (await result.ToJsonAsync()).MatchSnapshot();
        }
    }
}