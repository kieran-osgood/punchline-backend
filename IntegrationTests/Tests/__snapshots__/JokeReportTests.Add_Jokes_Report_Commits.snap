{
  "data": {
    "addJokeReport": {
      "jokeReport": {
        "id": "Sm9rZVJlcG9ydAppMQ==",
        "reportingUser": {
          "id": "VXNlcgppMQ==",
          "name": "Test User"
        },
        "reportedJoke": {
          "id": "Sm9rZQppMQ==",
          "title": "Cow With No Legs"
        }
      },
      "errors": null
    }
  }
}
